xtask/src/ui.rs

```diff
@@ -0,0 +1,480 @@
+use std::{env, fs};
+
+use crate::{temp, util, ImageTarget};
+use clap::Args;
+use cross::shell::MessageInfo;
+use cross::{docker, CommandExt, ToUtf8};
+
+#[derive(Args, Debug)]
+pub struct UiTest {
+    /// Provide verbose diagnostic output.
+    #[clap(short, long)]
+    pub verbose: bool,
+    /// Do not print cross log messages.
+    #[clap(short, long)]
+    pub quiet: bool,
+    /// Coloring: auto, always, never
+    #[clap(long)]
+    pub color: Option<String>,
+    /// Container engine (such as docker or podman).
+    #[clap(long)]
+    pub engine: Option<String>,
+    /// Specify a tag to use instead of the derived one, eg `local`
+    #[clap(long)]
+    pub tag: Option<String>,
+    /// Repository name for image.
+    #[clap(long, default_value = docker::CROSS_IMAGE)]
+    pub repository: String,
+    /// Do not pull the image (local image).
+    #[clap(long)]
+    pub no_pull: bool,
+    /// Target to run tests for.
+    #[clap(long)]
+    pub target: Option<ImageTarget>,
+}
+
+struct Utf8Output {
+    command: String,
+    status: ExitStatus,
+    stdout: String,
+    stderr: String,
+}
+
+impl Utf8Output {
+    fn bail(&self, message: &str) -> eyre::ErrReport {
+        eyre::eyre!(
+            r#"command unexpectedly {message} with exit status of {:?}:
+command
+=======
+{}
+
+stdout
+======
+{}
+
+stderr
+======
+{}
+"#,
+            self.status,
+            self.command,
+            self.stdout,
+            self.stderr
+        )
+    }
+
+    fn check(&self, message: &str, op: impl Fn(&Self) -> bool) -> cross::Result<()> {
+        match op(self) {
+            true => Ok(()),
+            false => Err(self.bail(message)),
+        }
+    }
+
+    fn success(&self) -> cross::Result<()> {
+        self.check("failed", |x| x.status.success())
+    }
+
+    fn failure(&self) -> cross::Result<()> {
+        self.check("succeeded", |x| !x.status.success())
+    }
+
+    fn no_fallback(&self) -> cross::Result<()> {
+        self.check("fallsback", |x| {
+            !x.stderr
+                .contains("[cross] note: Falling back to `cargo` on the host.")
+        })
+    }
+
+    fn fallsback(&self) -> cross::Result<()> {
+        self.check("does not fallback", |x| {
+            x.stderr
+                .contains("[cross] note: Falling back to `cargo` on the host.")
+        })
+    }
+
+    fn no_metadata(&self) -> cross::Result<()> {
+        self.check("has metadata", |x| {
+            x.stderr
+                .contains("[cross] warning: unable to get metadata for package")
+        })
+    }
+
+    fn no_stdout(&self) -> cross::Result<()> {
+        self.check("has stdout", |x| x.stdout.is_empty())
+    }
+
+    fn outside_package(&self) -> cross::Result<()> {
+        self.failure()?;
+        self.fallsback()?;
+        self.no_metadata()?;
+        self.no_stdout()?;
+        self.check("did not error", |x| {
+            x.stderr.contains("error: could not find `Cargo.toml`")
+        })
+    }
```

Comment:
Since we only run all these tests once, **most** of these can be moved out of here.